app/forms/providers/means/regular_transaction_form.rb

```diff
@@ -0,0 +1,158 @@
+
+        assign_regular_transaction_attributes
+
+        super
+      end
+
+      def save
+        return false unless valid?
+
+        ApplicationRecord.transaction do
+          destroy_transactions!
+
+          build_legal_aid_application_transaction_types
+          legal_aid_application.assign_attributes(legal_aid_application_attributes)
+          legal_aid_application.save!
+
+          regular_transactions.each(&:save!)
+        end
+
+        true
+      end
+
+      def transaction_type_ids=(ids)
+        @transaction_type_ids = if none_selected?
+                                  []
+                                else
+                                  [ids].flatten.compact_blank
+                                end
+      end
+
+      def transaction_type_options
+        TransactionType.not_children.active.where(operation: transaction_type_operation)
+      end
+
+      def none_selected
+        "none"
+      end
+
+    private
+
+      def transaction_type_operation
+        raise(
+          NotImplementedError,
+          "#transaction_type_operation must be implemented in the subclass",
+        )
+      end
+
+      def legal_aid_application_attributes
+        raise(
+          NotImplementedError,
+          "#legal_aid_application_attributes must be implemented in the subclass",
+        )
+      end
+
+      def none_selected?
+        @none_selected
+      end
+
+      def existing_transaction_type_ids
+        legal_aid_application
+          .transaction_types
+          .where(operation: transaction_type_operation)
+          .pluck(:id)
+      end
+
+      def assign_regular_transaction_attributes
+        regular_transactions.each do |transaction|
+          transaction_type = transaction.transaction_type
+          public_send("#{transaction_type.name}_amount=", transaction.amount)
+          public_send("#{transaction_type.name}_frequency=", transaction.frequency)
+        end
+      end
+
+      def build_legal_aid_application_transaction_types
+        transaction_type_ids.each do |transaction_type_id|
+          next if transaction_type_id.in?(legal_aid_application.transaction_type_ids)
+
+          legal_aid_application.legal_aid_application_transaction_types.build(
+            transaction_type_id:,
+          )
+        end
+      end
+
+      def destroy_transactions!
+        dependent_transaction_models.each do |model|
+          legal_aid_application
+            .public_send(model)
+            .includes(:transaction_type)
+            .where(transaction_type: { operation: transaction_type_operation })
+            .where.not(transaction_type_id: transaction_type_ids)
+            .destroy_all
+        end
+      end
+
+      def dependent_transaction_models
+        %i[
+          legal_aid_application_transaction_types
+          regular_transactions
+          cash_transactions
+        ]
+      end
```

Comment:
I think this highlights the awkwardness of the data modelling currently.

In an ideal world, we wouldn't need to explicitly destroy these records —it would be covered by `dependent: :destroy` declarations in ActiveRecord or `on_cascade: :delete` relationships defined in Postgres.

When I can, I'll spend a bit of time exploring how painful/less that change might be.

Reply:
Similarly, we might be able to compress some of the complexity around building/validating/persisting regular transaction records by defining nested attribute relationships —but I don't want to get ahead of myself just yet!